packages/runtime/container-runtime/src/garbageCollection.ts

```diff
@@ -865,59 +903,60 @@ export class GarbageCollector implements IGarbageCollector {
-     * garbage collector via `referenceAdded`.
-     * We validate that the references in the current reference graph should be a subset of the references in the last
-     * run's reference graph + references since the last run.
+     * Finds outbound references to datastores in the current GC graph that aren't in the previous GC graph and returns
+     * them if they are not added by GC's `addedOutboundReference`.
+     *
+     * In more simple terms:
+     * Missing Explicit References = Current References - Previous References - Explicitly Added References;
+     *
+     * Currently outbound references are references from a parent node to child datastore node
+     *
      * @param currentGCData - The GC data (reference graph) from the current GC run.
+     * @param previousGCData - The GC data (reference graph) from the previous GC run.
+     * @param explicitReferences - New references added by GC explicity between the previous and the current run.
+     * @returns - a list of missing explicit references
      */
-    private validateReferenceCorrectness(currentGCData: IGarbageCollectionData) {
+    private findMissingExplicitReferences(
+        currentGCData: IGarbageCollectionData,
+        previousGCData: IGarbageCollectionData,
+        explicitReferences: Map<string, string[]>,
+    ): IGCReference[] {
         assert(
-            this.gcDataFromLastRun !== undefined,
+            previousGCData !== undefined,
             0x2b7, /* "Can't validate correctness without GC data from last run" */
         );
 
-        // Get a list of all the outbound routes (or references) in the current GC data.
-        const currentReferences: string[] = [];
-        for (const [nodeId, outboundRoutes] of Object.entries(currentGCData.gcNodes)) {
-            /**
-             * Remove routes from a child node to its parent which is added implicitly by the runtime. For instance,
-             * each adds its data store as an outbound route to mark it as referenced if the DDS is referenced.
-             * We won't get any explicit notification for these references so they must be removed before validation.
-             */
-            const explicitRoutes = outboundRoutes.filter((route) => !nodeId.startsWith(route));
-            currentReferences.push(...explicitRoutes);
-        }
-
-        // Get a list of outbound routes (or references) from the last run's GC data plus references added since the
-        // last run that were notified via `referenceAdded`.
-        const explicitReferences: string[] = [];
-        for (const [, outboundRoutes] of Object.entries(this.gcDataFromLastRun.gcNodes)) {
-            explicitReferences.push(...outboundRoutes);
-        }
-        this.referencesSinceLastRun.forEach((outboundRoutes: string[]) => {
-            explicitReferences.push(...outboundRoutes);
-        });
+        const getOutboundRoutes = (nodeId: string, routes: string[]): string[] => {
```

Comment:
This is only used in convertGraphToOutounndReferences and so making this a function does not add any value. You can make this inline in convertGraphToOutounndReferences.